sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java

```diff
@@ -96,6 +102,13 @@ private static DoFnSignature parseSignature(Class<? extends DoFn<?, ?>> fnClass)
     }
     errors.checkNotNull(inputT, "Unable to determine input type");
 
+    // Find the state and timer declarations in advance of validating
+    // method parameter lists
+    Map<String, TimerDeclaration> timerDeclarations = analyzeTimerDeclarations(errors, fnClass);
```

Comment:
Nit: probably better to have state-related things first, and timer-related things second ("state and timers" are part of the same proposal, but I wouldn't say that they are so tightly coupled conceptually that they have to always go together).